tests/end_to_end/end_to_end_lib.py

```diff
@@ -0,0 +1,275 @@
+        server_name = master_server_name(i)
+        config_file_content += indent + indent + server_name + ": 0.0.0.0\n"
+    for i in range(1, num_of_chunk_server+1):
+        server_name = chunk_server_name(i)
+        config_file_content += indent + indent + server_name + ": 0.0.0.0\n"
+
+    # Generate a few configuration params
+    config_file_content += "disk:\n"
+    config_file_content += indent + "block_size_mb: " + str(block_size_mb) \
+                               + "\n"
+    config_file_content += indent + "min_free_disk_space_mb: " \
+                               + str(min_free_disk_space_mb) + "\n"
+
+    config_file_content += "timeout:\n"
+    config_file_content += indent + "grpc: " + str(grpc_timeout_s) + "s\n"
+    config_file_content += indent + "lease: " + str(lease_timeout_s) + "s\n"
+    config_file_content += indent + "heartbeat: " + str(heartbeat_timeout_s) \
+                               + "s\n"
+    config_file_content += indent + "client_cache: " \
+                               + str(client_cache_timeout_m) + "m\n"
+
+    # Write to the config file
+    with open(config_filename, "w") as config_file:
+        config_file.write(config_file_content)
+
+    # Return the configuration data
+    return config_data
+
+# Start a cluster which is composed by a number of master and chunk servers
+# as specified by the configurable params. Optionally this function takes a path
+# to the directory to store the logs. Return the processes in a list.
+def start_master_and_chunk_servers(config_filename, log_directory = None,
+        num_of_master_server = 1, num_of_chunk_server = 3, block_size_mb = 64,
+        min_free_disk_space_mb = 100, grpc_timeout_s = 10, lease_timeout_s = 60,
+        heartbeat_timeout_s = 30, client_cache_timeout_m = 10):
+    # Generate the config file and config data
+    config_data = generate_config_and_config_file(config_filename, 
+                      num_of_master_server, num_of_chunk_server, block_size_mb,
+                      min_free_disk_space_mb, grpc_timeout_s, lease_timeout_s,
+                      heartbeat_timeout_s, client_cache_timeout_m)
+
+    master_and_chunk_server_procs = []
+    for master_server_name in config_data["servers"]["master_servers"]:
+        # Specify the command for master node
+        command = [master_server_binary(), "--config_path=%s"%config_filename,
+                       "--master_name=%s"%master_server_name]
+        master_proc = None
+        if log_directory != None:
+            os.makedirs(log_directory, exist_ok=True)
+            log_file = open(log_directory + "/" + master_server_name \
+                                + ".txt", "w+")
+            master_proc = subprocess.Popen(command, stderr=log_file)
+        else:
+            master_proc = subprocess.Popen(command)
+        
+        master_and_chunk_server_procs.append(master_proc)
+    
+    # TODO(Xi): Add chunk server's launching once it is shaped up
```

Comment:
chunk server `run` binary doesn't do anything right now, but listen on the port, so it doesn't hurt to include it right?

Reply:
Yes, sounds good, let me add that.